### ISSUE TITLE
Possibly wonky permissions set on MySQL Unix socket file in `/var/lib/mysql/<app_name>/mysql.sock`

### ISSUE THREAD

**Issue by NucciTheBoss** [canonical/mysql-router-operator#20]
## Steps to reproduce

1. `git clone git@github.com:canonical/slurmdbd-operator.git`
2. `tox -e integration`
3. Wait a few minutes for test deployment to stabilize. Integration tests will pass because slurmdbd has _StorageHost_ set to 127.0.0.1 and _StoragePort_ set to 3306, so we will need to go inside the slurmdbd/0 unit to configure it to use the mysql socket instead.
4. `juju ssh slurmdbd/0`
5. `sudo systemctl stop slurmdbd mysqlrouter`
6. `sudo echo 'MYSQL_UNIX_PORT="/var/lib/mysql/slurmdbd/mysql.sock"'>> /etc/default/slurmdbd`
7. Comment out _StorageHost_ and _StoragePort_ configuration parameters in slurmdbd configuration file. Comments are made by prepending the line with "#": `sudo nano /etc/slurm/slurmdbd.conf`
8. `sudo systemctl start mysqlrouter`
9. `sudo systemctl start slurmdbd`
10. Check if slurmdbd port is open and communication on: `sudo lsof -t -n -iTCP:6819 -sTCP:LISTEN`

After step 9 you will notice that port 6819 (the port slurmdbd uses to communicate with slurmctld) is not open. This means that slurmdbd is not functional; `systemctl status slurmdbd` will only tell you if the slurmdbd daemon was able to start successfully.

## Expected behavior

The expected behavior is that slurmdbd should successfully be able to use the Unix socket file `/var/lib/mysql/slurmdbd/mysql.sock` to communicate with the MySQL database on mysql/0. slurmdbd defaults to using the socket file if the _StorageHost_ and _StoragePort_ configuration parameters are not specified.

## Actual behavior

slurmdbd fails to connect to the MySQL database on mysql/0:

```text
[2023-02-22T15:46:13.524] accounting_storage/as_mysql: _check_mysql_concat_is_sane: MySQL server version is: 8.0.32-router
[2023-02-22T15:46:13.528] error: Database settings not recommended values: innodb_buffer_pool_size innodb_lock_wait_timeout
[2023-02-22T15:46:13.610] error: mysql_real_connect failed: 2002 Can't connect to local MySQL server through socket '/var/lib/mysql/slurmdbd/mysql.sock' (13)
[2023-02-22T15:46:13.611] error: Problem getting cache of data
[2023-02-22T15:46:13.611] error: mysql_real_connect failed: 2002 Can't connect to local MySQL server through socket '/var/lib/mysql/slurmdbd/mysql.sock' (13)
[2023-02-22T15:46:13.611] error: unable to re-connect to as_mysql database
```

## Versions

Operating system: Ubuntu 20.04.5 LTS

Juju CLI: 3.1.0-genericlinux-amd64

Juju agent: 3.1.0

Charm revision: Revision 60 from channel dpe/edge

LXD: 5.11

## Log output

`juju debug-log --replay` does not have any helpful output as this is an issue more pertaining to how the slurmdbd charm interacts with MySQL router outside of relations. The DatabaseCreatedEvent works as expected.

Output of `sudo ls -la /var/lib/mysql/slurmdbd/mysql`:

```text
ubuntu@juju-c1e60b-2:~$ sudo ls -la /var/lib/mysql/slurmdbd
total 11
drwx------ 5 mysql mysql   14 Feb 22 15:07 .
drwxr-xr-x 3 mysql mysql    3 Feb 22 15:07 ..
drwx------ 2 mysql mysql    8 Feb 22 15:07 data
drwx------ 2 mysql mysql    3 Feb 22 15:07 log
srwxrwxrwx 1 mysql mysql    0 Feb 22 15:07 mysql.sock
srwxrwxrwx 1 mysql mysql    0 Feb 22 15:07 mysqlro.sock
-rw------- 1 mysql mysql 2335 Feb 22 15:07 mysqlrouter.conf
-rw------- 1 mysql mysql   94 Feb 22 15:07 mysqlrouter.key
-rw-r--r-- 1 mysql mysql    5 Feb 22 15:07 mysqlrouter.pid
srwxrwxrwx 1 mysql mysql    0 Feb 22 15:07 mysqlx.sock
srwxrwxrwx 1 mysql mysql    0 Feb 22 15:07 mysqlxro.sock
drwx------ 2 mysql mysql    2 Feb 22 15:07 run
-rwx------ 1 mysql mysql  293 Feb 22 15:07 start.sh
-rwx------ 1 mysql mysql  179 Feb 22 15:07 stop.sh
```

Output of `ps ax | grep slurmdbd`:

```text
ubuntu@juju-c1e60b-2:~$ ps ax | grep slurmdbd
  10236 ?        Sl     0:01 /usr/bin/mysqlrouter -c /var/lib/mysql/slurmdbd/mysqlrouter.conf
  13332 ?        Ssl    0:00 /usr/sbin/slurmdbd -D -s
  13340 pts/1    S+     0:00 grep --color=auto slurmdbd
```

## Additional context

Changing the permissions on `/var/lib/mysql` and `/var/lib/mysql/slurmdbd`, and restarting mysqlrouter and slurmdbd fixes the issue of slurmdbd not being able to connect to MySQL through the Unix socket file:

```shell
sudo chmod 755 /var/lib/mysql /var/lib/mysql/slurmdbd
sudo systemctl restart mysqlrouter
sudo systemctl restart slurmdbd 
```

**Comment by paulomach** [comment]
Thanks for the in depth report @NucciTheBoss , we will try to fit in next pulse.

**Comment by shayancanonical** [comment]
@NucciTheBoss I was able to test your slurmdbd fork with the mysql and mysql-router charms after the fix in PR #29. I had some issues with the `version` file missing in the slurmdbd charm, but I temporarily commented out the line that expects the file for the sake of the test. Additionally, as you suggested above, I had to add `MYSQL_UNIX_PORT` as a config to slurmdbd and comment out the `StorageHost` and `StoragePort` in the `slurmdbd.conf`. Validated that after these changes, slurmdbd starts up:

```
[2023-03-16T19:40:03.825] accounting_storage/as_mysql: _check_mysql_concat_is_sane: MySQL server version is: 8.0.32-router
[2023-03-16T19:40:03.860] error: Database settings not recommended values: innodb_buffer_pool_size innodb_lock_wait_timeout
[2023-03-16T19:40:04.081] slurmdbd version 21.08.8-2 started
```

**Comment by NucciTheBoss** [comment]
Excellent! Thank you for this work!

> I was able to test your slurmdbd fork with the mysql and mysql-router charms after the fix in PR #29. I had some issues with the `version` file missing in the slurmdbd charm.

Yes, this is a known issue with all the slurm charms currently. The CI/CD pipeline we put in place automatically generates the version file so it is not an issue when folks pull slurm charms from Charmhub. I have an open bug report for addressing the issue, be we just have other priorities at the moment :sweat_smile:

Thank you again for fixing this!
